File: app-flutter/lib/features/content/cache/content_metadata_cache.dart
import 'package:flutter/foundation.dart';
import 'package:hive/hive.dart';
import 'package:nebras_mobile_app/core/data/content_ordering.dart';
import 'package:nebras_mobile_app/features/content/model/content_model.dart';

part 'content_metadata_cache.g.dart';

/// نسخة Hive خفيفة من ميتادات المحتوى فقط. لا نخزن ملفات التشغيل هنا؛
/// الهدف أن تبقى العناوين والأوصاف والصور المصغّرة متاحة عند انقطاع الشبكة.
@HiveType(typeId: 3)
class ContentMetadataCacheEntry {
  @HiveField(0)
  final String id;
  @HiveField(1)
  final String title;
  @HiveField(2)
  final String author;
  @HiveField(3)
  final String description;
  @HiveField(4)
  final String type;
  @HiveField(5)
  final String thumbnailUrl;
  @HiveField(6)
  final String? sourceUrl;
  @HiveField(7)
  final int? sizeInBytes;
  @HiveField(8)
  final String section;
  @HiveField(9)
  final String? subSection;
  @HiveField(10)
  final String? sectionName;
  @HiveField(11)
  final int createdAtMs;
  @HiveField(12)
  final int updatedAtMs;
  @HiveField(13)
  final String createdBy;
  @HiveField(14)
  final int cachedAtMs;

  const ContentMetadataCacheEntry({
    required this.id,
    required this.title,
    required this.author,
    required this.description,
    required this.type,
    required this.thumbnailUrl,
    required this.section,
    required this.createdAtMs,
    required this.updatedAtMs,
    required this.createdBy,
    required this.cachedAtMs,
    this.sourceUrl,
    this.sizeInBytes,
    this.subSection,
    this.sectionName,
  });

  factory ContentMetadataCacheEntry.fromContent(Content content) {
    return ContentMetadataCacheEntry(
      id: content.id,
      title: content.title,
      author: content.author,
      description: content.description,
      type: content.type.name,
      thumbnailUrl: content.thumbnailUrl,
      sourceUrl: content.sourceUrl,
      sizeInBytes: content.sizeInBytes,
      section: content.section,
      subSection: content.subSection,
      sectionName: content.sectionName,
      createdAtMs: content.createdAt.millisecondsSinceEpoch,
      updatedAtMs: content.updatedAt.millisecondsSinceEpoch,
      createdBy: content.createdBy,
      cachedAtMs: DateTime.now().millisecondsSinceEpoch,
    );
  }

  Content toContent() {
    return Content(
      id: id,
      title: title,
      author: author,
      description: description,
      type: parseContentType(type),
      thumbnailUrl: thumbnailUrl,
      sourceUrl: sourceUrl,
      sizeInBytes: sizeInBytes,
      section: section,
      subSection: subSection,
      sectionName: sectionName,
      createdAt: DateTime.fromMillisecondsSinceEpoch(createdAtMs),
      updatedAt: DateTime.fromMillisecondsSinceEpoch(updatedAtMs),
      createdBy: createdBy,
    );
  }
}

class ContentMetadataCache extends ChangeNotifier {
  ContentMetadataCache(this._box);

  final Box<ContentMetadataCacheEntry> _box;
  final Set<String> _offlinePreviewIds = <String>{};

  List<Content> getAll() {
    final items = _box.values.map((entry) => entry.toContent()).toList();
    sortContentOldestFirst(items);
    return items;
  }

  Content? getById(String id) => _box.get(id)?.toContent();

  bool isOfflinePreview(String id) => _offlinePreviewIds.contains(id);

  Future<void> upsertAll(Iterable<Content> contents) async {
    final rows = <String, ContentMetadataCacheEntry>{};
    for (final content in contents) {
      final id = content.id.trim();
      if (id.isEmpty) continue;
      rows[id] = ContentMetadataCacheEntry.fromContent(content);
    }
    if (rows.isEmpty) return;
    await _box.putAll(rows);
    notifyListeners();
  }

  /// نخزن لقطة حيّة ونزيل عنها وسم "معاينة أوفلاين". لا نحذف كل الكاش هنا
  /// لأن الصفحة قد تكون محدودة pagination؛ واجهة المستخدم الحية نفسها هي
  /// التي تنظف المحذوفات فور عودة الاتصال.
  Future<void> cacheLiveItems(Iterable<Content> contents) async {
    final ids = contents
        .map((content) => content.id.trim())
        .where((id) => id.isNotEmpty)
        .toSet();
    await upsertAll(contents);
    _offlinePreviewIds.removeAll(ids);
    notifyListeners();
  }

  Future<void> reconcileWithLiveIds(Iterable<String> liveIds) async {
    final keep = liveIds
        .map((id) => id.trim())
        .where((id) => id.isNotEmpty)
        .toSet();
    final staleKeys = _box.keys
        .whereType<String>()
        .where((id) => !keep.contains(id))
        .toList(growable: false);
    if (staleKeys.isNotEmpty) {
      await _box.deleteAll(staleKeys);
      notifyListeners();
    }
  }

  bool hasMetadata(String id) => _box.containsKey(id);

  bool has(String id) => hasMetadata(id);

  Future<void> cacheMany(Iterable<Content> contents) => upsertAll(contents);

  /// عند فشل الشبكة نعرض الميتادات المحفوظة فقط، ونوسمها كي تُعطّل أزرار
  /// تشغيل الوسائط غير المحمّلة مع بقاء النص والصورة المصغرة قابلين للتصفح.
  List<Content> offlineContents() {
    final contents = getAll();
    _offlinePreviewIds
      ..clear()
      ..addAll(contents.map((content) => content.id));
    return contents;
  }

  static bool isOfflinePreviewId(String id) {
    return id.startsWith('offline_preview:');
  }

  Content markOfflinePreview(Content content) {
    final id = content.id.trim();
    _offlinePreviewIds.add(id);
    return Content(
      id: isOfflinePreviewId(id) ? id : 'offline_preview:$id',
      title: content.title,
      author: content.author,
      description: content.description,
      type: content.type,
      thumbnailUrl: content.thumbnailUrl,
      sourceUrl: content.sourceUrl,
      sizeInBytes: content.sizeInBytes,
      section: content.section,
      subSection: content.subSection,
      sectionName: content.sectionName,
      createdAt: content.createdAt,
      updatedAt: content.updatedAt,
      createdBy: content.createdBy,
    );
  }

  List<HomeSectionCacheGroup> offlineGroups() {
    final grouped = <String, List<Content>>{};
    for (final content in offlineContents()) {
      final key = content.section.trim().isNotEmpty
          ? content.section.trim()
          : 'offline_metadata';
      grouped.putIfAbsent(key, () => <Content>[]).add(content);
    }
    return grouped.entries
        .map((entry) {
          final first = entry.value.first;
          final title = (first.sectionName ?? '').trim().isNotEmpty
              ? first.sectionName!.trim()
              : entry.key;
          return HomeSectionCacheGroup(
            id: 'offline:${entry.key}',
            title: title,
            items: entry.value,
          );
        })
        .toList(growable: false);
  }

  static List<HomeSectionCacheGroup> offlineSectionsFromEntries(
    Iterable<ContentMetadataCacheEntry> entries,
  ) {
    final grouped = <String, List<Content>>{};
    for (final entry in entries) {
      final content = entry.toContent();
      final key = content.section.trim().isNotEmpty
          ? content.section.trim()
          : 'offline_metadata';
      grouped.putIfAbsent(key, () => <Content>[]).add(content);
    }
    return grouped.entries
        .map((entry) {
          final first = entry.value.first;
          final title = (first.sectionName ?? '').trim().isNotEmpty
              ? first.sectionName!.trim()
              : entry.key;
          return HomeSectionCacheGroup(
            id: 'offline:${entry.key}',
            title: title,
            items: entry.value,
          );
        })
        .toList(growable: false);
  }
}

extension OfflineContentPreview on Content {
  Content asOfflinePreview() {
    final previewId = ContentMetadataCache.isOfflinePreviewId(id)
        ? id
        : 'offline_preview:$id';
    return Content(
      id: previewId,
      title: title,
      author: author,
      description: description,
      type: type,
      thumbnailUrl: thumbnailUrl,
      sourceUrl: sourceUrl,
      sizeInBytes: sizeInBytes,
      section: section,
      subSection: subSection,
      sectionName: sectionName,
      createdAt: createdAt,
      updatedAt: updatedAt,
      createdBy: createdBy,
    );
  }
}

class OfflineContentPolicy {
  const OfflineContentPolicy._();

  static bool canOpen(Content content, {required bool hasLocalFile}) {
    if (!ContentMetadataCache.isOfflinePreviewId(content.id)) return true;
    if (content.type != ContentType.video &&
        content.type != ContentType.youtube &&
        content.type != ContentType.audio) {
      return true;
    }
    return hasLocalFile;
  }
}

class HomeSectionCacheGroup {
  final String id;
  final String title;
  final List<Content> items;

  const HomeSectionCacheGroup({
    required this.id,
    required this.title,
    required this.items,
  });
}

File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/model/ContentMetadataCache.kt
package com.nebras.mobile.core.model

import android.util.Log
import com.nebras.mobile.core.data.LocalStore
import com.nebras.mobile.core.data.LocalStoreKeys
import com.nebras.mobile.core.data.sortedContentOldestFirst
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject
import java.util.Date
import java.util.concurrent.ConcurrentHashMap

/**
 * صفّ خفيف من ميتادات المحتوى فقط (بديل `ContentMetadataCacheEntry` في Hive).
 * لا نخزّن ملفّات التشغيل هنا؛ الهدف أن تبقى العناوين والأوصاف والصور
 * المصغّرة متاحة عند انقطاع الشبكة.
 */
data class ContentMetadataCacheEntry(
    val id: String,
    val title: String,
    val author: String,
    val description: String,
    val type: String,
    val thumbnailUrl: String,
    val sourceUrl: String? = null,
    val sizeInBytes: Int? = null,
    val section: String,
    val subSection: String? = null,
    val sectionName: String? = null,
    val createdAtMs: Long,
    val updatedAtMs: Long,
    val createdBy: String,
    val cachedAtMs: Long,
) {
    fun toContent(): Content = Content(
        id = id,
        title = title,
        author = author,
        description = description,
        type = parseContentType(type),
        thumbnailUrl = thumbnailUrl,
        sourceUrl = sourceUrl,
        sizeInBytes = sizeInBytes,
        section = section,
        subSection = subSection,
        sectionName = sectionName,
        createdAt = Date(createdAtMs),
        updatedAt = Date(updatedAtMs),
        createdBy = createdBy,
    )

    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id)
        put("title", title)
        put("author", author)
        put("description", description)
        put("type", type)
        put("thumbnailUrl", thumbnailUrl)
        sourceUrl?.let { put("sourceUrl", it) }
        sizeInBytes?.let { put("sizeInBytes", it) }
        put("section", section)
        subSection?.let { put("subSection", it) }
        sectionName?.let { put("sectionName", it) }
        put("createdAtMs", createdAtMs)
        put("updatedAtMs", updatedAtMs)
        put("createdBy", createdBy)
        put("cachedAtMs", cachedAtMs)
    }

    companion object {
        fun fromContent(content: Content): ContentMetadataCacheEntry =
            ContentMetadataCacheEntry(
                id = content.id,
                title = content.title,
                author = content.author,
                description = content.description,
                type = content.type.name,
                thumbnailUrl = content.thumbnailUrl,
                sourceUrl = content.sourceUrl,
                sizeInBytes = content.sizeInBytes,
                section = content.section,
                subSection = content.subSection,
                sectionName = content.sectionName,
                createdAtMs = content.createdAt.time,
                updatedAtMs = content.updatedAt.time,
                createdBy = content.createdBy,
                cachedAtMs = System.currentTimeMillis(),
            )

        fun fromJson(json: JSONObject): ContentMetadataCacheEntry =
            ContentMetadataCacheEntry(
                id = json.optString("id"),
                title = json.optString("title"),
                author = json.optString("author"),
                description = json.optString("description"),
                type = json.optString("type"),
                thumbnailUrl = json.optString("thumbnailUrl"),
                sourceUrl = json.optString("sourceUrl").takeIf { it.isNotEmpty() },
                sizeInBytes = if (json.has("sizeInBytes")) json.optInt("sizeInBytes") else null,
                section = json.optString("section"),
                subSection = json.optString("subSection").takeIf { it.isNotEmpty() },
                sectionName = json.optString("sectionName").takeIf { it.isNotEmpty() },
                createdAtMs = json.optLong("createdAtMs"),
                updatedAtMs = json.optLong("updatedAtMs"),
                createdBy = json.optString("createdBy"),
                cachedAtMs = json.optLong("cachedAtMs"),
            )
    }
}

/**
 * كاش الميتادات — بديل صندوق Hive `content_metadata_cache` فوق [LocalStore].
 * الصفّ يحتفظ بالخريطة في الذاكرة ويكتبها كاملةً عند كل تعديل (حجمها
 * ميتادات نصّيّة فقط، والكتابة `apply()` غير حاجبة).
 */
class ContentMetadataCache(private val store: LocalStore) {

    private val entries = ConcurrentHashMap<String, ContentMetadataCacheEntry>()
    private val offlinePreviewIds = java.util.Collections.newSetFromMap(
        ConcurrentHashMap<String, Boolean>(),
    )

    private val _revision = MutableStateFlow(0L)

    /** يتغيّر عند كل تعديل — بديل `notifyListeners()`. */
    val revision: StateFlow<Long> = _revision.asStateFlow()

    init {
        load()
    }

    private fun load() {
        val root = store.getJsonObject(LocalStoreKeys.BOX_CONTENT_METADATA_CACHE) ?: return
        for (key in root.keys()) {
            val obj = root.optJSONObject(key) ?: continue
            runCatching { ContentMetadataCacheEntry.fromJson(obj) }
                .onSuccess { entries[key] = it }
        }
    }

    private fun persist() {
        runCatching {
            val root = JSONObject()
            entries.forEach { (key, entry) -> root.put(key, entry.toJson()) }
            store.putJsonObject(LocalStoreKeys.BOX_CONTENT_METADATA_CACHE, root)
        }
        _revision.value = _revision.value + 1
    }

    fun getAll(): List<Content> =
        entries.values.map { it.toContent() }.sortedContentOldestFirst()

    fun getById(id: String): Content? = entries[id]?.toContent()

    fun isOfflinePreview(id: String): Boolean = offlinePreviewIds.contains(id)

    fun upsertAll(contents: Iterable<Content>) {
        var changed = false
        for (content in contents) {
            val id = content.id.trim()
            if (id.isEmpty()) continue
            entries[id] = ContentMetadataCacheEntry.fromContent(content)
            changed = true
        }
        if (changed) persist()
    }

    /**
     * نخزّن لقطة حيّة ونزيل عنها وسم «معاينة أوفلاين». لا نحذف كلّ الكاش هنا
     * لأنّ الصفحة قد تكون محدودة pagination؛ الواجهة الحيّة نفسها هي التي
     * تنظّف المحذوفات فور عودة الاتصال.
     */
    fun cacheLiveItems(contents: Iterable<Content>) {
        val ids = contents.map { it.id.trim() }.filter { it.isNotEmpty() }.toSet()
        upsertAll(contents)
        offlinePreviewIds.removeAll(ids)
    }

    /**
     * يحذف من الكاش العناصر التي لم تَعُد في اللقطة الحيّة (محتوى حذفته
     * اللوحة) كي لا يبقى ظاهراً أوفلاين.
     *
     * 🛡️ **حارسان ضدّ اختفاء المحتوى نهائياً**: لقطة Firestore قد تصل
     * **مبتورة** (`.limit(500)`) أو **جزئيّة** لحظة انقطاع. الاعتماد عليها
     * لحذف كلّ ما ليس فيها كان يمسح الكتالوج المحفوظ بالكامل:
     *   1) لا نحذف شيئاً إذا كانت اللقطة فارغة.
     *   2) لا نحذف دفعة كبيرة (> ثلث الكاش) — مؤشّر لقطة مبتورة/جزئيّة.
     */
    fun reconcileWithLiveIds(liveIds: Iterable<String>) {
        val keep = liveIds.map { it.trim() }.filter { it.isNotEmpty() }.toSet()
        if (keep.isEmpty()) return

        val total = entries.size
        val staleKeys = entries.keys.filter { !keep.contains(it) }
        if (staleKeys.isEmpty()) return

        if (total > 0 && staleKeys.size > total / 3) {
            Log.d(
                TAG,
                "reconcile skipped — would prune ${staleKeys.size}/$total " +
                    "(suspected truncated/partial snapshot)",
            )
            return
        }

        staleKeys.forEach(entries::remove)
        persist()
    }

    fun hasMetadata(id: String): Boolean = entries.containsKey(id)

    fun has(id: String): Boolean = hasMetadata(id)

    fun cacheMany(contents: Iterable<Content>) = upsertAll(contents)

    /**
     * عند فشل الشبكة نعرض الميتادات المحفوظة فقط ونوسمها كي تُعطَّل أزرار
     * تشغيل الوسائط غير المحمّلة مع بقاء النصّ والصورة قابلَين للتصفّح.
     */
    fun offlineContents(): List<Content> {
        val contents = getAll()
        offlinePreviewIds.clear()
        offlinePreviewIds.addAll(contents.map { it.id })
        return contents
    }

    fun markOfflinePreview(content: Content): Content {
        val id = content.id.trim()
        offlinePreviewIds.add(id)
        return content.copy(
            id = if (isOfflinePreviewId(id)) id else "$OFFLINE_PREVIEW_PREFIX$id",
        )
    }

    fun offlineGroups(): List<HomeSectionCacheGroup> = groupBySection(offlineContents())

    companion object {
        private const val TAG = "ContentMetadataCache"
        private const val OFFLINE_PREVIEW_PREFIX = "offline_preview:"

        fun isOfflinePreviewId(id: String): Boolean = id.startsWith(OFFLINE_PREVIEW_PREFIX)

        fun offlineSectionsFromEntries(
            entries: Iterable<ContentMetadataCacheEntry>,
        ): List<HomeSectionCacheGroup> = groupBySection(entries.map { it.toContent() })

        private fun groupBySection(contents: List<Content>): List<HomeSectionCacheGroup> {
            val grouped = LinkedHashMap<String, MutableList<Content>>()
            for (content in contents) {
                val key = content.section.trim().ifEmpty { "offline_metadata" }
                grouped.getOrPut(key) { mutableListOf() }.add(content)
            }
            return grouped.map { (key, items) ->
                val first = items.first()
                val title = first.sectionName?.trim()?.takeIf { it.isNotEmpty() } ?: key
                HomeSectionCacheGroup(id = "offline:$key", title = title, items = items)
            }
        }
    }
}

/** يحوّل عنصراً إلى «معاينة أوفلاين» بمعرّف موسوم. */
fun Content.asOfflinePreview(): Content {
    val previewId = if (ContentMetadataCache.isOfflinePreviewId(id)) {
        id
    } else {
        "offline_preview:$id"
    }
    return copy(id = previewId)
}

object OfflineContentPolicy {
    fun canOpen(content: Content, hasLocalFile: Boolean): Boolean {
        if (!ContentMetadataCache.isOfflinePreviewId(content.id)) return true
        if (content.type != ContentType.VIDEO &&
            content.type != ContentType.YOUTUBE &&
            content.type != ContentType.AUDIO
        ) {
            return true
        }
        return hasLocalFile
    }
}

data class HomeSectionCacheGroup(
    val id: String,
    val title: String,
    val items: List<Content>,
)
